### PR TITLE
ENH: Support mask in duplicated algorithm

### PR DESCRIPTION
**Pull request by phofl**
- [x] xref #45776 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jorisvandenbossche 

Benchmarks don't show any changes for duplicated.

```
ser = pd.Series([1, 2, pd.NA] * 1_000_000, dtype="Int64")
%timeit ser.duplicated(keep="first")
# 7.22 ms ± 42.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
ser = pd.Series((list(range(1, 1_000_000)) + [pd.NA]) * 3, dtype="Int64")
%timeit ser.duplicated(keep="first")
# 12.3 ms ± 402 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

### CODE REVIEW

File: pandas/core/algorithms.py
Comment (phofl) [inline pandas/core/algorithms.py:1044]: This assumes, that we don't want to add duplicated to the ea interface

**Comment by phofl** [comment]
@mroeschke would you mind having a look?

File: pandas/_libs/hashtable_func_helper.pxi.in
Comment (mroeschke) [inline pandas/_libs/hashtable_func_helper.pxi.in:169]: Would be nice if we could template `for i in ...` based on `{{ if keep == "first"/"last" }}` but could be a follow up.

File: pandas/core/algorithms.py
Comment (mroeschke) [reply pandas/core/algorithms.py:1044]: Might be good to add an issue to discuss this. Kinda unfortunate we need this exception for `BasedMaskedDtype`

File: pandas/_libs/hashtable_func_helper.pxi.in
Comment (mroeschke) [inline pandas/_libs/hashtable_func_helper.pxi.in:189]: So this `if/elif` block is when `keep=False`? I am having troubling understanding why we need to track first 2 NAs instead of just assigning `out[i] = 1`

Comment (phofl) [reply pandas/_libs/hashtable_func_helper.pxi.in:189]: Yes correct. When keep=False we can only set True when we have multiple nas, so we have to set the first na to true after we saw a second one

Comment (mroeschke) [reply pandas/_libs/hashtable_func_helper.pxi.in:189]: Oh duh, right. Need at least 2 NAs to be a duplicate 🤦

**Review by mroeschke — COMMENTED** [review]
Do we need any additional testing for this?

File: pandas/core/algorithms.py
Comment (phofl) [reply pandas/core/algorithms.py:1044]: Yeah not happy about this either, will open a follow up issue to discuss

File: pandas/_libs/hashtable_func_helper.pxi.in
Comment (phofl) [reply pandas/_libs/hashtable_func_helper.pxi.in:169]: Oh good idea, would do as a follow up to keep diff reviewable.

**Comment by phofl** [comment]
Good point, assumed we had some but could not finde them 😂 will add some for all three cases

**Comment by jorisvandenbossche** [comment]
This is a much bigger change, but from looking at this PR, I started wondering: should we see multiple NAs as duplicated, or should we rather propagate NAs? 
In the past, propagating was never an option, since we didn't have a nullable boolean dtype. But similarly as NAs now propagate in comparison ops, they could also propagate here?

**Comment by mroeschke** [comment]
> I started wondering: should we see multiple NAs as duplicated, or should we rather propagate NAs?

IMO if I was using `duplicated`, I would be a little more surprised If NAs persisted instead of indicating whether the row with NA was a duplicate with another row with NA (in the same position). I get though that comparison semantics with NA should propagate NA.

I think a natural follow up question with propagating NAs would be "How can i check whether this row with NA is a duplicate of this row with NA?"

**Comment by jorisvandenbossche** [comment]
Ah, good point, I wasn't considering the DataFrame case where you are checking for duplicated _rows_ (multiple columns at once), but I was only thinking about the more simple Series.duplicated (in which case the position doesn't matter, it's just which values are NA in that Series).

Yes, for DataFrames it certainly makes sense to consider NAs at the same location as equal, when considering whether two rows are equal.

**Comment by phofl** [comment]
Added the tests

I think this is a topic worth having a discussion, from a users perspective it would be a bit confusing If we consider them equal in one case and unequal in another case.

Would this impact operations like isin and merges and similar?

**Comment by phofl** [comment]
@mroeschke are you ok with merging this as is?

File: pandas/tests/series/methods/test_duplicated.py
Comment (mroeschke) [inline pandas/tests/series/methods/test_duplicated.py:62]: Nit: Could you add one more `NA` at the end? IIRC I think technically the last

```
                    else:
                        out[i] = 1
```
Doesn't have coverage with this test

Comment (phofl) [reply pandas/tests/series/methods/test_duplicated.py]: Good point, thx. Added

**Comment by mroeschke** [comment]
Small comment otherwise LGTM. Generally I am in favor of not propagating NA for  both `DataFrame/Series.duplicated` as this PR does. Is it worth discussing that in a separate issue @jorisvandenbossche before merging?

**Review by mroeschke — APPROVED** [review]

File: pandas/tests/series/methods/test_duplicated.py
Comment (jorisvandenbossche) [reply pandas/tests/series/methods/test_duplicated.py]: But we should probably also cover the case when `out[i]` does not get set, i.e. there is only a single NA?

**Review by jorisvandenbossche — APPROVED** [review]
Small request for an additional test with non-duplicate NA, otherwise looks good!

**Comment by jorisvandenbossche** [comment]
> Is it worth discussing that in a separate issue @jorisvandenbossche before merging?

Given that this PR just preserves the current behaviour, that certainly doesn't need to block this PR from being merged.

If it would only be for Series.duplicated, I would take the time to open an issue to argue for propagating NAs, but since it's less clear how to deal with the DataFrame.duplicated case, I don't feel strongly enough about it.

File: pandas/tests/series/methods/test_duplicated.py
Comment (phofl) [reply pandas/tests/series/methods/test_duplicated.py]: thx, this found a bug

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**
